### PR TITLE
Only look up package.json type if module is node16/nodenext or file is in node_modules

### LINKED ISSUE #50647: Typescript [4.8.2] is adding invalid javascript for *.cjs files
# Bug Report

My project is set to emit ESNext modules. I have a variety of TS files, but I also have a single `.cjs` file, which needs to remain CommonJS to try and require legacy packages.

My input looks like this:

```js
module.exports = {
  tryRequire(modulePath) {
    return require(modulePath);
  },
};
```

I expect the output I get from TypeScript 4.7.4:

```js
module.exports = {
  tryRequire(modulePath) {
    return require(modulePath);
  },
};
```

When I upgrade to TypeScript 4.8.2, it now emits an extra `export {}` at the end of the .cjs file:

```js
module.exports = {
  tryRequire(modulePath) {
    return require(modulePath);
  },
};
export {}
```

This causes Node to fail parsing the `.cjs` file because `export` does not exist in the .cjs context.

### 🕗 Version & Regression Information

Worked in 4.7.4, broken in 4.8.2, also broken in current nightly (4.9.0-dev.20220905).

### Repro here

https://github.com/dzearing/ts-repro-cjs

### LINKED ISSUE #54752: ESM-only packages come with synthesized default in `moduleResolution: bundler`
# Bug Report

### 🔎 Search Terms

module resolution bundler synthesized default

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&moduleResolution=100#code/PTAEAEFsHsBMFcA2BTASsgztR8AuBLaAOwC5QAjeI2FAJwCgQIYEUzMjkAPXRscXAENaAc2S52GTj3r18kAA7RauUJFiCMuADSgA3qFzQAyrlr4iI0AF9QAM1rRIoAETrNuALR58iT0c8tc0sXAG5QJlwATwVMMgBmADoAJkSABll3LUSjU2CRelyzCwL6AGNiLTUNLWTQAF5QQQB3QXxVeSUVAAo3Gq8fPwCgkpcASnDImLjQJNSM+izcVNhkO0EkXiWVtY3EXByTYstF-tSi-NluLtU9ayA)

### 💻 Code

```ts
// @moduleResolution: bundler
// @module: esnext
// @target: esnext

import mdast, { toString } from "mdast-util-to-string"; // types: 3.2.0

mdast.toString
toString

const mdast2 = await import("mdast-util-to-string"); // types: 3.2.0

mdast2.default
mdast2.default.toString
mdast2.toString

export {}
```

### 🙁 Actual behavior

`default` export is available here and points to the whole module. I don't think this is right, even with `moduleResolution: bundler` but I can be wrong about this since it's really hard to know what kind of assumptions and goals this mode has when it comes to subtle details

### 🙂 Expected behavior

I don't expect `default` to be available here since this ESM-only package **doesn't** have a default export. 

Note that even if this isn't a bug, `.default` available on that dynamically imported binding should behave in the same way in the IDE as the specifier coming from the import statement. Right now the hover info isn't available on it and go to (type?) definition doesn't work on it with cmd+click.

cc @andrewbranch

### PR DESCRIPTION
**Pull request by andrewbranch**
This walks back the heuristics added to non-Node.js module modes in #57896 a bit, based mostly on feedback in #58663. In 5.5-beta, an explicit package.json `"type"` or module-format-specific file extension started overriding `"module": "commonjs"` or `"module": "esnext"` in tsconfig.json. This impacted both type checking (relevant primarily for imports of `node_modules` dependencies, but with some effects on local files, e.g. with `verbatimModuleSyntax`) and emit (relevant for local project source files, if emit was enabled). The feedback in #58663 was that the effects on local files were unwelcome because the emitted files (whether from tsc or another tool) were going to be placed in some other context tsc didn’t know about, so the package.json tsc was finding during compilation shouldn’t be considered during the build.

With this PR, we only look up package.json `"type"` if `--module` is `node16`/`nodenext` (as we’ve always done) _or_ if the file is inside `node_modules`.

~~Some testing and experimentation is still needed to ensure this doesn’t create inconsistencies with symlinked monorepos.~~ This doesn’t seem to cause inconsistencies with workspaces-style monorepos, because module resolution takes the realpath of files in node_modules before we see if we need to look up a package.json `"type"`. So in a project references example like

```
.
└── packages/
    ├── app/
    │   ├── node_modules/shared ↪︎
    │   ├── tsconfig.json
    │   ├── package.json
    │   └── index.ts
    └── shared/
        ├── dist/index.d.ts
        ├── src/index.ts
        ├── tsconfig.json
        └── package.json
```

When we build `shared/tsconfig.json`, we decide that `shared/src/index.ts` is **not inside node_modules**. Then, when we build `app/tsconfig.json`, even though we resolve `app/node_modules/shared/dist/index.d.ts`, we realpath it to `shared/dist/index.d.ts` and count it as **not inside node_modules**. Having this consistency between input and output files regardless of how they’re resolved is what matters most. The caveat is that the realpath and project references logic can cause the type checking of `app/tsconfig.json` during development to behave differently than it would after publishing/deploying into a situation where `app/node_modules/shared` is no longer a symlink, but a real node_modules package.

Fixes #54752
Fixes #50647

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot test top400

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by andrewbranch** [comment]
Perf should be unchanged or better, but

@typescript-bot perf test faster

**Review by weswigham — COMMENTED** [review]
We don't need these new `ModuleSpecifierResolutionHost` members to be public because we automatically attach them or have fallbacks, right? Because these'll be missing from a user-provided `host` object. We probably wannt mark 'em as optional and handle appropriately if they're non-public.

**Review by sheetalkamat — APPROVED** [review]

**Comment by andrewbranch** [comment]
> We don't need these new `ModuleSpecifierResolutionHost` members to be public

@weswigham `ModuleSpecifierResolutionHost` itself is internal.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -47012,10 +47028,22 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
+                    moduleKind === ModuleKind.Preserve &&
+                    node.kind !== SyntaxKind.ImportEqualsDeclaration &&
+                    node.kind !== SyntaxKind.VariableDeclaration &&
+                    host.getEmitModuleFormatOfFile(getSourceFileOfNode(node)) === ModuleKind.CommonJS
+                ) {
+                    // In `--module preserve`, ESM input syntax emits ESM output syntax, but there will be times
+                    // when we look at the `impliedNodeFormat` of this file and decide it's CommonJS. To avoid
+                    // that inconsistency, we disallow ESM syntax in files that are unambiguously CommonJS in
+                    // this mode.
+                    error(node, Diagnostics.ESM_syntax_is_not_allowed_in_a_CommonJS_module_when_module_is_set_to_preserve);
```
Comment (jakebailey) [inline src/compiler/checker.ts:47045]: Just to be clera here, this is explicitly about `.cjs` files containing ESM syntax? Or will `.js` files error too when `type=commonjs` is set? The latter seems like it'd be a problem for bun/tsc misc scripts, etc, or projects that set that explicitly for completeness but then bundle.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -41,6 +40,7 @@ import {
     canHaveDecorators,
     canHaveLocals,
     canHaveModifiers,
+    type CanHaveModuleSpecifier,
```
Comment (jakebailey) [inline src/compiler/utilities.ts:43]:
```suggestion
    CanHaveModuleSpecifier,
```

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -8685,11 +8705,13 @@ function isFileModuleFromUsingJSXTag(file: SourceFile): Node | undefined {
...
  * in SourceFile construction.
  */
-function isFileForcedToBeModuleByFormat(file: SourceFile): true | undefined {
+function isFileForcedToBeModuleByFormat(file: SourceFile, options: CompilerOptions): true | undefined {
     // Excludes declaration files - they still require an explicit `export {}` or the like
     // for back compat purposes. The only non-declaration files _not_ forced to be a module are `.js` files
     // that aren't esm-mode (meaning not in a `type: module` scope).
-    return (file.impliedNodeFormat === ModuleKind.ESNext || (fileExtensionIsOneOf(file.fileName, [Extension.Cjs, Extension.Cts, Extension.Mjs, Extension.Mts]))) && !file.isDeclarationFile ? true : undefined;
+    //
+    // TODO: extension check never considered compilerOptions; should impliedNodeFormat?
```
Comment (jakebailey) [inline src/compiler/utilities.ts:8713]: Is this TODO still relevant?

File: src/compiler/checker.ts (hunk 1 above)
Comment (andrewbranch) [reply src/compiler/checker.ts:47045]: It’s now just about `.cjs` files.

Comment (jakebailey) [reply src/compiler/checker.ts:47045]: Thanks for clarifying!

**Comment by andrewbranch** [comment]
There’s a knip failure but it’s public API?

**Comment by weswigham** [comment]
> ModuleSpecifierResolutionHost itself is internal.

The specifics are internal, yes, but we expose and consume public objects and interfaces that we use as and expect to be `ModuleSpecifierResolutionHost`s which cannot be correctly implemented without these being made public in some way. There's a definite reason almost everything on `ModuleSpecifierResolutionHost` is optional. For example, `Program` has a second (`@internal`) declaration declaring it extends `ModuleSpecifierResolutionHost`. If there are any 3rd-party `Program`s (like our LS or builder construct internally to orchestrate builds, but external), they won't know they need these new methods on their `Program` to satisfy `ModuleSpecifierResolutionHost` uses, and since they're *required*, we'll just crash at runtime when they're inevitably missing.

**Comment by andrewbranch** [comment]
There are 43 non-optional internal methods and properties declared on public `Program` interface declaration—we haven’t made it possible to implement a third-party `Program` up from its public interface.

**Comment by andrewbranch** [comment]
@weswigham any other feedback on the API (or any of the PR)? I’d like to get this in soon.

**Pull request merged**